packages/arb-token-bridge-ui/src/components/App/App.tsx

```diff
@@ -143,7 +142,71 @@ const AppContent = (): JSX.Element => {
 const Injector = ({ children }: { children: React.ReactNode }): JSX.Element => {
   const actions = useActions()
 
-  const [globalBridge, setGlobalBridge] = useState<Bridge | null>(null)
+  const [globalBridgeData, setGlobalBridgeData] =
+    useState<BridgeContextInterface>({
+      bridge: null,
+      walletType: null
+    })
+
+  const setGlobalBridge = async (bridge: Bridge | null) => {
+    if (!bridge) {
+      setGlobalBridgeData({
+        bridge: null,
+        walletType: null
+      })
+    } else {
+      const { l1Signer, l2Signer, l1Provider, l2Provider } = bridge
+      const l1Address = await l1Signer.getAddress()
+      const l2Address = await l2Signer.getAddress()
+      const l1AddressIsSmartContract =
+        (await l1Provider.getCode(l1Address)).length > 2
+      const l2AddressIsSmartContract =
+        (await l2Provider.getCode(l2Address)).length > 2
+
+      const walletType: WalletType = await (async () => {
```

Comment:
I think a lot of this logic in the IIFE could be instead in an arb-ts module. This L1 / L2 wallet smart contract inference could be useful for others

Reply:
agreed on this and the other arb-ts suggestion; gonna save for the more general refactor